File: sources/CSharp/src/Ers/Visualization/InstancedModel.cs
using Ers.Engine;
using System.Numerics;

namespace Ers
{
    /// <summary>
    /// A model of which multiple instances are drawn in one draw call.
    /// </summary>
    public class InstancedModel : IDisposable
    {
        /// <summary>
        /// Native pointer to the core instance.
        /// </summary>
        public IntPtr CorePtr;

        /// <summary>
        /// Construct a new empty <see cref="InstancedModel"/>.
        /// </summary>
        public InstancedModel() { CorePtr = ErsEngine.ERS_InstancedModel_Create(); }

        /// <summary>
        /// Construct a new <see cref="InstancedModel"/> with a given mesh.
        /// </summary>
        /// <param name="mesh"></param>
        public InstancedModel(Mesh mesh)
        {
            CorePtr = ErsEngine.ERS_InstancedModel_Create();
            ErsEngine.ERS_InstancedModel_SetMesh(CorePtr, mesh.CorePtr);
        }

        internal InstancedModel(IntPtr corePtr) { CorePtr = corePtr; }

        /// <summary>
        /// Set the mesh used for instancing.
        /// </summary>
        /// <param name="mesh"></param>
        public void SetMesh(Mesh mesh) => ErsEngine.ERS_InstancedModel_SetMesh(CorePtr, mesh.CorePtr);

        /// <summary>
        /// Finalizer.
        /// </summary>
        ~InstancedModel() => DisposeInner();

        /// <summary>
        /// Indicate the InstancedModel is no longer required and can be freed from memory.
        /// </summary>
        public void Dispose()
        {
            DisposeInner();
            GC.SuppressFinalize(this);
        }

        private void DisposeInner()
        {
            if (CorePtr != IntPtr.Zero)
            {
                ErsEngine.ERS_InstancedModel_Destroy(CorePtr);
                CorePtr = IntPtr.Zero;
            }
        }

        /// <summary>
        /// Clear all instanced.
        ///
        /// <para>Call this every frame so instances don't spill over.</para>
        /// </summary>
        public void Clear() => ErsEngine.ERS_InstancedModel_Clear(CorePtr);

        /// <summary>
        /// Push a new instance of the instanced model to be rendered.
        /// </summary>
        /// <param name="position">The position of the instance.</param>
        /// <param name="rotation">The rotation of the instance in turns (1 turn = 360 degrees).</param>
        /// <param name="scale">The scale of the instance.</param>
        public void PushInstance(Vector3 position, Vector3 rotation = default, Vector3 scale = default)
        {
            if (rotation == default)
                rotation = Vector3.Zero;
            if (scale == default)
                scale = Vector3.One;

            ErsEngine.ERS_InstancedModel_PushInstance(
                CorePtr, position.X, position.Y, position.Z, rotation.X, rotation.Y, rotation.Z, scale.X, scale.Y, scale.Z);
        }
    }
}

File: sources/CPP/include/Ers/Visualization/InstancedModel.cpp
#include "InstancedModel.h"

#include "Ers/Api.h"
#include "Ers/Math/HMM/VectorMath.h"

namespace Ers
{
    InstancedModel::InstancedModel()
    {
        corePtr = Ers::Engine::ERS_InstancedModel_Create();
    }

    InstancedModel::InstancedModel(Mesh& mesh)
    {
        corePtr = Ers::Engine::ERS_InstancedModel_Create();
        Ers::Engine::ERS_InstancedModel_SetMesh(corePtr, mesh.CorePtr());
    }

    InstancedModel::~InstancedModel()
    {
        Ers::Engine::ERS_InstancedModel_Destroy(CorePtr());
    }

    void InstancedModel::SetMesh(Mesh& mesh)
    {
        Ers::Engine::ERS_InstancedModel_SetMesh(CorePtr(), mesh.CorePtr());
    }

    void InstancedModel::PushInstance(const TransformComponent& globalTransform)
    {
        PushInstance(globalTransform.GetGlobalPosition(), globalTransform.GetGlobalRotation(), globalTransform.GetGlobalScale());
    }

    void InstancedModel::PushInstance(Vector3 pos, Vector3 rotation, Vector3 scale)
    {
        Ers::Engine::ERS_InstancedModel_PushInstance(
            CorePtr(), pos.X, pos.Y, pos.Z, rotation.X, rotation.Y, rotation.Z, scale.X, scale.Y, scale.Z);
    }

    void InstancedModel::Clear()
    {
        Ers::Engine::ERS_InstancedModel_Clear(CorePtr());
    }

    void* InstancedModel::CorePtr()
    {
        return corePtr;
    }

    const void* const InstancedModel::CorePtr() const
    {
        return corePtr;
    }
} // namespace Ers
